src/dbnode/storage/repair_test.go

```diff
@@ -675,11 +695,131 @@ func TestDatabaseRepairPrioritizationLogic(t *testing.T) {
+			repairState: repairStatesByNs{
+				"ns2": namespaceRepairStateByTime{
+					flushTimeEndNano: repairState{
+						Status:      repairSuccess,
+						LastAttempt: time.Time{},
+					},
+				},
+			},
+			expectedNS1Repairs: []expectedRepair{
+				{
+					xtime.Range{Start: flushTimeEnd, End: flushTimeEnd.Add(blockSize)},
+					errors.New("ns1 repair error"),
+				},
+				{
+					xtime.Range{Start: flushTimeStart, End: flushTimeStart.Add(blockSize)},
+					nil,
+				},
+			},
+			expectedNS2Repairs: []expectedRepair{
+				{
+					xtime.Range{Start: flushTimeStart, End: flushTimeStart.Add(blockSize)},
+					errors.New("ns2 repair error"),
+				},
+				{
+					xtime.Range{Start: flushTimeEnd, End: flushTimeEnd.Add(blockSize)},
+					nil,
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.title, func(t *testing.T) {
+			opts := DefaultTestOptions().SetRepairOptions(testRepairOptions(ctrl))
+			mockDatabase := NewMockdatabase(ctrl)
+
+			databaseRepairer, err := newDatabaseRepairer(mockDatabase, opts)
+			require.NoError(t, err)
+			repairer := databaseRepairer.(*dbRepairer)
+			repairer.nowFn = func() time.Time {
+				return now
+			}
+			if tc.repairState == nil {
+				tc.repairState = repairStatesByNs{}
+			}
+			repairer.repairStatesByNs = tc.repairState
+
+			mockDatabase.EXPECT().IsBootstrapped().Return(true)
+
+			var (
+				ns1        = NewMockdatabaseNamespace(ctrl)
+				ns2        = NewMockdatabaseNamespace(ctrl)
+				namespaces = []databaseNamespace{ns1, ns2}
+			)
+			ns1.EXPECT().Options().Return(nsOpts).AnyTimes()
+			ns2.EXPECT().Options().Return(nsOpts).AnyTimes()
+
+			ns1.EXPECT().ID().Return(ident.StringID("ns1")).AnyTimes()
+			ns2.EXPECT().ID().Return(ident.StringID("ns2")).AnyTimes()
+
+			//Setup expected ns1 repair invocations for each repaired time range
+			var ns1RepairExpectations = make([]*gomock.Call, len(tc.expectedNS1Repairs))
+			for i, ns1Repair := range tc.expectedNS1Repairs {
+				ns1RepairExpectations[i] = ns1.EXPECT().
+					Repair(gomock.Any(), ns1Repair.expectedRepairRange).
+					Return(ns1Repair.mockRepairResult)
+			}
+			gomock.InOrder(ns1RepairExpectations...)
+
+			//Setup expected ns2 repair invocations for each repaired time range
+			var ns2RepairExpectations = make([]*gomock.Call, len(tc.expectedNS2Repairs))
+			for i, ns2Repair := range tc.expectedNS2Repairs {
+				ns2RepairExpectations[i] = ns2.EXPECT().
+					Repair(gomock.Any(), ns2Repair.expectedRepairRange).
+					Return(ns2Repair.mockRepairResult)
+			}
+			gomock.InOrder(ns2RepairExpectations...)
+
+			mockDatabase.EXPECT().OwnedNamespaces().Return(namespaces, nil)
+
+			require.NotNil(t, repairer.Repair())
```

Comment:
Failures already being logged here - https://github.com/m3db/m3/blob/58297189116ce3f13e5306fcc4cab0a910cfe456/src/dbnode/storage/repair.go#L669

I was wondering the same thing but then I realized namespace which get repaired on second call is redundant because namespaces which get repaired is only a function of repairState in the test setup. So instead of calling repair twice we can just set repairState to different values before calling Repair(). Thats what I have tried to do with setting ns2's repair state differently from ns1.